include/circt/Dialect/Synth/Transforms/CutRewriter.h

```diff
@@ -0,0 +1,267 @@
+  Cut() = default;
+
+  /// Get the root operation of this cut.
+  /// The root operation produces the output of the cut.
+  mlir::Operation *getRoot() const;
+
+  void dump(llvm::raw_ostream &os) const;
+
+  /// Merge this cut with another cut to form a new cut.
+  /// The new cut combines the operations from both cuts with the given root.
+  Cut mergeWith(const Cut &other, Operation *root) const;
+  Cut reRoot(Operation *root) const;
+
+  /// Get the number of inputs to this cut.
+  unsigned getInputSize() const;
+
+  /// Get the number of operations in this cut.
+  unsigned getCutSize() const;
+
+  /// Get the number of outputs from root operation.
+  unsigned getOutputSize() const;
+
+  /// Get the truth table for this cut.
+  /// The truth table represents the boolean function computed by this cut.
+  const BinaryTruthTable &getTruthTable() const;
+
+  /// Get the NPN canonical form for this cut.
+  /// This is used for efficient pattern matching against library components.
+  const NPNClass &getNPNClass() const;
+
+  /// Get the permutated inputs for this cut based on the given pattern NPN.
+  void getPermutatedInputs(const NPNClass &patternNPN,
+                           SmallVectorImpl<Value> &permuatedInputs) const;
+};
+
+/// Represents a cut that has been successfully matched to a rewriting pattern.
+///
+/// This class encapsulates the result of matching a cut against a rewriting
+/// pattern during optimization. It stores the matched pattern, the
+/// cut that was matched, and timing information needed for optimization.
+/// TODO: Currently only added as a placeholder.
+class MatchedPattern {
+public:
+  /// Default constructor creates an invalid matched pattern.
+  MatchedPattern() = default;
+};
+
+/// Manages a collection of cuts for a single logic node using priority cuts
+/// algorithm.
+///
+/// Each node in the combinational logic network can have multiple cuts
+/// representing different ways to group it with surrounding logic. The CutSet
+/// manages these cuts and selects the best one based on the optimization
+/// strategy (area or timing).
+///
+/// The priority cuts algorithm maintains a bounded set of the most promising
+/// cuts to avoid exponential explosion while ensuring good optimization
+/// results.
+class CutSet {
+private:
+  llvm::SmallVector<Cut, 4> cuts; ///< Collection of cuts for this node
+  std::optional<MatchedPattern> matchedPattern; ///< Best matched pattern found
+  bool isFrozen = false; ///< Whether cut set is finalized
+
+public:
+  /// Finalize the cut set by removing duplicates and selecting the best
+  /// pattern.
+  ///
+  /// This method:
+  /// 1. Removes duplicate cuts based on inputs and root operation
+  /// 2. Limits the number of cuts to prevent exponential growth
+  /// 3. Matches each cut against available patterns
+  /// 4. Selects the best pattern based on the optimization strategy
+  void finalize(const CutRewriterOptions &options);
```

Comment:
How does this function have access to the available patterns? Is this going to be a separate argument to the function later, or do the cuts or cut sets have some reference to a common object that holds match patterns?

Reply:
Yes, good question and sorry that it was not clear from comments that I changed the signature to split the PR. 
In the follow up it will be changed to take a lambda which will be passed from `CutRewriter` class that holds all the available patterns.  
 
https://github.com/uenoku/circt/blob/4e6d3e0b80745b221b3da3a9893a1303386ad198/lib/Synthesis/CutRewriter.cpp#L421
https://github.com/uenoku/circt/blob/4e6d3e0b80745b221b3da3a9893a1303386ad198/lib/Synthesis/CutRewriter.cpp#L590-L592